sdks/python/apache_beam/transforms/distribution_counter.pyx

```diff
@@ -0,0 +1,134 @@
+    element >>= 1
+  return bit_count
+
+
+cdef int64_t get_log10_round_to_floor(int64_t element):
+  cdef int64_t number_of_leading_zeros = 64 - bit_length(element)
+  cdef int64_t y = MAX_LONG_10_FOR_LEADING_ZEROS[number_of_leading_zeros]
+  return y - compare_to(element, POWER_TEN[y])
+
+
+cdef class DistributionAccumulator(object):
+  """Distribution Counter:
+  contains value distribution statistics and methods for incrementing
+  Attributes:
+    min: minimum value of all inputs.
+    max: maximum value of all inputs.
+    count: total count of all inputs.
+    sum: sum of all inputs.
+    first_bucket_offset: starting index of the first stored bucket.
+    last_bucket_offset: end index of buckets.
+    buckets: histogram buckets of value counts for a distribution
+                                                    (1,2,5 bucketing).
+             max bucket_index is 58( sys.maxint as input)
+    buckets_per_10: 3 buckets for every power of ten -> 1, 2, 5
+  """
+  def __init__(self):
+    self.min = INT64_MAX
+    self.max = 0
+    self.count = 0
+    self.sum = 0
+    self.first_bucket_offset = 58
+    self.last_bucket_offset = 0
+    self.buckets = <int64_t*> malloc(sizeof(int64_t) * 59)
+    memset(self.buckets, 0, sizeof(int64_t) * 59)
+    self.buckets_per_10 = 3
+
+  cdef bint add_input(self, int64_t element) except -1:
+    if element < 0:
+      raise ValueError('Distribution counters support only non-negative value')
+    self.min = min(self.min, element)
+    self.max = max(self.max, element)
+    self.count += 1
+    self.sum += element
+    cdef int64_t bucket_index = self.calculate_bucket_index(element)
+    self.buckets[bucket_index] += 1
+    self.first_bucket_offset = min(self.first_bucket_offset, bucket_index)
+    self.last_bucket_offset = max(self.last_bucket_offset, bucket_index)
+
+  cdef int64_t calculate_bucket_index(self, int64_t element):
+    """Calculate the bucket index for the given element"""
+    if element == 0:
+      return 0
+    cdef int64_t log10_floor = get_log10_round_to_floor(element)
+    cdef int64_t power_of_ten = POWER_TEN[log10_floor]
+    cdef int64_t bucket_offset = 0
+    if element <  power_of_ten<<1:
+      bucket_offset = 0
+    elif element < ((power_of_ten<<2) + power_of_ten):
+      bucket_offset = 1
+    else:
+      bucket_offset = 2
+    return 1 + log10_floor * self.buckets_per_10 + bucket_offset
+
+  cpdef object get_add_input_fn(self):
+    """Get add_input function"""
+    return self.add_input
+
+  cpdef object translate_to_histogram(self, histogram):
+    """Translate buckets into Histogram"""
+    histogram.firstBucketOffset = self.first_bucket_offset
+    histogram.bucketCounts = []
+    cdef int64_t index = self.first_bucket_offset
+    for index in range(self.first_bucket_offset, self.last_bucket_offset+1):
+      histogram.bucketCounts.append(self.buckets[index])
+
+  cpdef bint add_inputs_for_test(self, elements) except -1:
+    """ONLY used for unit tests"""
```

Comment:
<!--new_thread; commit:e0fdd73db41481933d3173aabe6300b65003abd6; resolved:0-->
Why not implement these in the test file, if they are only needed for tests?

Reply:
Since test files are python codes and they cannot access to cdef function. However, we need cdef add_input() to make operation faster. Then these test_purpose_only cpdef functions will behave as a bridge to make cdef functions callable from python codes.

Reply:
Are you saying that defining `add_input` as `cpdef` as opposed to `cdef` makes this significantly slower?

Reply:
yes, we had a test before. cdef can reduce almost 60-70ns per element, compared to cpdef.

Reply:
Okay, thanks for the clarification.  Can you define, for example, `cdef _fast_add_input()` and use it in the Cython code here, and also define `cpdef add_input()` which just wraps this and does `return self._fast_add_input(...)`?  Similarly for `calculate_bucket_index`.  Could you also add a comment that this pattern is done for performance?

Reply:
No, what I mean is that you can provide an external interface with `cpdef add_input`, but you can continue using the current fast `cdef _fast_add_input` directly in the Cython code.

Reply:
Means, do following signature changes:
```
cdef add_input -> cdef _fast_add_input
cpdef get_add_input_fn -> cpdef add_input
```
with the same implementations ?

Reply:
add_inputs_for_test will only be called from microbenchmark and unit tests. And we only care about performance when running microbenchmark. So performance can be calculated more precisely by calling add_inputs_for_test once, then a bunch of add_input will be called through c call.

Reply:
You could put add_inputs_for_test in a compiled test file, but it seems overkill to create a whole new file just for this one function, so I think it's fine to leave it here.